### DISCUSSION TITLE
[API Request] update/changed timestamp value on all entities

### DISCUSSION THREAD

**Discussion by dbrader** [opensubsonic/open-subsonic-api#32] [API requests]
### Type of change

API Clarification

### Proposal description

A change timestamp should be added to all entities so that remote clients can determine if their local cache is current without further action.

A change timestamp attribute should also be added to the top level list entity to indicate if any entities in the list have changed.  The client can then skip iterating through the list if nothing has changed.

getPlaylists already contains the change attribute for this very reason but it was not added to the other entities.

For example, in the getStarred2 api.  A remote client needs to iterate through the albums listed and subsequently call getAlbum to determine if the songs in the album have changed.

getStarred2 returns

```
<starred2>
	<album id="2044" name="The Division Bell" artist="Pink Floyd" artistId="415" coverArt="al-2044" songCount="11" duration="3986" playCount="27" created="2012-02-15T07:16:49.000Z" starred="2023-01-02T12:43:10.300Z" year="1994" genre="Rock"/>
	<album id="1994" name="Rhythm of the Saints" artist="Paul Simon" artistId="404" coverArt="al-1994" songCount="10" duration="2680" playCount="0" created="2010-08-29T08:52:57.000Z" starred="2022-11-01T05:04:30.232Z" year="1990" genre="Blues"/>
	<album id="1987" name="Graceland" artist="Paul Simon" artistId="404" coverArt="al-1987" songCount="22" duration="5173" playCount="180" created="2015-11-20T13:54:17.000Z" starred="2022-11-01T05:04:24.300Z" year="1986" genre="Blues"/>
	<album id="2370" name="Aja" artist="Steely Dan" artistId="498" coverArt="al-2370" songCount="7" duration="2398" playCount="191" created="2012-02-15T01:50:14.000Z" starred="2022-08-09T18:49:41.932Z" year="1972" genre="Soft Rock"/>
	<song id="34126" parent="34114" isDir="false" title="Africa" album="Toto IV" artist="Toto" track="10" year="1982" genre="AOR Classic Rock" coverArt="34114" size="5291484" contentType="audio/mpeg" suffix="mp3" duration="263" bitRate="160" path="Toto/Toto IV/10 Africa.mp3" playCount="35" discNumber="1" created="2012-02-15T06:19:39.000Z" starred="2023-01-17T04:00:42.176Z" albumId="2881" artistId="602" type="music"/>
	<song id="42295" parent="42177" isDir="false" title="Moneytalks" album="1990-The Razor's Edge" artist="Ac/dc" track="3" year="1990" genre="Hard Rock" coverArt="42177" size="9046770" contentType="audio/mpeg" suffix="mp3" duration="226" bitRate="320" path="AC-DC/1 Studio albums/1990 The Razor's Edge (Australian Editions) @320/03 Moneytalks.mp3" playCount="35" created="2022-07-06T16:14:18.316Z" starred="2022-08-09T18:48:14.349Z" albumId="3762" artistId="740" type="music"/>
	<song id="17779" parent="17702" isDir="false" title="All My Love" album="In Through The Out Door" artist="Led Zeppelin" track="6" year="1979" genre="Hard Rock" coverArt="17702" size="11403392" contentType="audio/mpeg" suffix="mp3" duration="356" bitRate="256" path="Led Zeppelin/In Through The Out Door/06 All My Love.mp3" playCount="34" created="2022-06-01T05:14:39.101Z" starred="2022-08-09T18:48:10.207Z" albumId="1543" artistId="327" type="music"/>
</starred2>
```

### Backward compatibility impact

adding as attribute to entity would not break compatibility

### Backward compatibility

- [X] No backward compatibility impact.

### API details

All entities should contain a "changed" attribute containing a timestamp to reduce network traffic and processing time

Example:

<album id="1987" name="Graceland" artist="Paul Simon" artistId="404" coverArt="al-1987" songCount="22" duration="5173" playCount="180" created="2015-11-20T13:54:17.000Z" starred="2022-11-01T05:04:24.300Z" year="1986" genre="Blues" **changed="2023-04-11T13:54:17.000Z"**/>

### Security impacts

none

### Alternative solutions

My current solution is to iterate through the entity lists, when the changed value is not available, but this obviously causes more network traffic and overhead.

**Comment by Tolriq** [comment]
+5000 :) But needs the rest of the API (Sorting and filtering on it) to be really useful and start to offer differential sync to clients.

There's also the complexity of knowing the kind of changes that happens. Typically the changedValue can't indicate that a song is removed from the database. Or artists ...

There's also indirect changes like an album is changed, but the change is a new artist, you then need to also sync that new artist. And doing differential on every entry to detect those can be highly inefficient.

So it to resume it would be nice to extend this to allow proper differential sync for offline first clients.

**Reply by dbrader** [reply]
The main purpose of this api change would be to improve steady state sync time/efficient.  While sorting and filtering are a nice to have most offline first clients would do that locally anyhow.

The indirect changes (like artist) most likely would not affect offline first clients since they would be syncing all artists in a similar manner to albums and playlists etc.  So when a change is detected the client would write over the entity in its entirety.  Conflict detection would be overkill and just add complexity the user doesn't care about imho.

**Reply by Tolriq** [reply]
I'm not sure we are on the same page here. 

>While sorting and filtering are a nice to have most offline first clients would do that locally anyhow.

This is for the client to be able to see everything changed after a date to be able to detect and sync those changes.

Having albums returning a changed date without a way to filter on that would mean that the client still needs to get all the album to know the changed ones, that's highly inefficient. Same for songs (Even more without a proper API to list all songs and requiring to use the search3 endpoint without any ordering / filtering).

> The indirect changes (like artist) most likely would not affect offline first clients since they would be syncing all artists in a similar manner to albums and playlists etc. So when a change is detected the client would write over the entity in its entirety. Conflict detection would be overkill and just add complexity the user doesn't care about imho.

The indirect change needs to be detectable to ease the sync. If you change the artist of an album and only detect the album change you sync the album that will have an artistId that is unknown to the client, so the client will need to sync that artist. Do detect that it means that for each album / songs synced it needs to check if the ids are known or not, hence the conflict handling and hence why having a way to know changes incompatible with differential sync are important, so that clients can do a full sync, because yes handling all the conflict detection is overkill and add complexity.

You do not have any repository or info what client / server do you represent?

**Reply by dbrader** [reply]
Yes, I misunderstood.  You are absolutely right filtering/sorting would be a very nice additional feature.  

In this situation, the client would be syncing the artist list as well.  Any additions or changes to artists would be picked up from the artist list sync.  You just need to update the referencing while syncing album/song.

I am the developer of the GoSONIC Android client.

**Reply by dbrader** [reply]
@Tolriq For the filtered output, how does a client know which entities were deleted?

**Reply by Tolriq** [reply]
They can't it requires a new endpoint. When working on the Kodi API a libraryStatus endpoint was added that returns all the necessary info to decide and allow differential syncs.

```
      "librarylastupdated": { "type": "string" },
      "librarylastcleaned": { "type": "string" },
      "artistlinksupdated": { "type": "string" },
      "songslastadded": { "type": "string" },
      "albumslastadded": { "type": "string" },
      "artistslastadded": { "type": "string" },
      "genreslastadded": { "type": "string" },
      "songsmodified": { "type": "string" },
      "albumsmodified": { "type": "string" },
      "artistsmodified": { "type": "string" }
```

It would be nice to have that too. So added/changed to everything + Filter / sort by them. With that offline first client can do differential sync and greatly improve end user experience.

**Reply by dbrader** [reply]
@Tolriq What would the client use the filtered output for?  Seems redundant if you need to process the differential output to get the current state.

**Reply by Tolriq** [reply]
To get the actual data to sync ....
App last sync was at X. Call the status endpoint.
Changes not compatible with diff sync do a full sync.
Changes compatible then get all the data changed after X
..

Basic of differential sync. This is important to be able to sync Play count, last played, favorites, ... All the data that change frequently but can be synced fast.

**Reply by dbrader** [reply]
Not sure what you mean by changes *not* compatible with diff sync vs changes compatible with diff sync.

**Reply by Tolriq** [reply]
Well it's in the name, this is not really the place to discuss client side architecture.

Anyway as you have probably noticed there's relations between things, songs / albums / artists. The relation is based on IDs.
Any change that does not touches the IDs can be easily synced via filtering on last changed since there's nothing that can be broken.

As soon as anything touches IDs like additions / deletions the situation becomes more complex.

There's basically 2 ways to address things depending on the server.
Either the server offers full journalisation of database changes and you can replay them locally. This is not the case here and too complex to ask for all servers to handle.
Or you need to know what changes can be easily synced like only additions, and those who can't like deletions.
With current API and filtering any deletion would be by definition not compatible with differential sync.

**Reply by dbrader** [reply]
I think this is the place to discuss this since we are talking about which approach we should use for improved syncing.

I've worked on both types of systems.  Those systems offering journalling and RESTful systems that focus on entity state.

Systems that cause lots of changes are not necessarily a good fit for journaling since they create lots of entries in the change table and playing those changes back causes a lot of overhead.  Especially when the same entity is changed multiple times or there are long delays between sync.  Journaling is good for huge datasets where a small portion of the data is changing.

So a complete song list for instance might be a good fit for journaling, maybe.

Still not sure why you would choose to playback the journal to pickup deletions and a few other things then have to go off and get changes since x.  Wouldn't the journaling contain the changes since x as well?

After working on both systems, as a general rule, I'm in favor of a more RESTful approach where we simply offer the current state at the time of sync.  It's far too easy for the journal to become too noisy and cause sync headaches. It's also much less prone to catastrophic bugs and easier to implement.

Maybe I've just been bitten by journaling issues where the entire sync falls flat on its face due to iterating through too many journal entries or a missing entry causing database inconsistency. ;)

**Reply by Tolriq** [reply]
> So a complete song list for instance might be a good fit for journaling, maybe.

You are mixing things here it seems. Journaling is at the global database level if we want proper state and relationships. This sentence does not make a lot of sense in this concept.

> Still not sure why you would choose to playback the journal to pickup deletions and a few other things then have to go off and get changes since x. Wouldn't the journaling contain the changes since x as well?

Again this is not what I said, the journaling contains all the changes. Journaling is done at server side, knowing at what X time the client did sync is known by the client it then handle the journal actions after that X. Having the server do journaling by clients would be insane to ask to the servers.

> I'm in favor of a more RESTful approach where we simply offer the current state at the time of sync

Does not fit at all the requirement for differential syncs and optimized syncs. If you only expose the state at instant X this means that clients needs to get all the data so can only do complete sync.....
Having a client sync 300 000 songs just to get a dozens of last played or 2 favorites changes is not good.

We are talking about offline first apps that have a copy of the data and all the relationships. Those clients needs to know the changes that they can sync safely with differential sync (So asking for the data changed after the last date of sync) and know if there's changes that prevent that since that date and trigger a full sync.

Having an endpoint exposing the last changes to differentiate, metadata/user data change, addition and deletions / relationship changes allows clients to decide what they do.

Having the created/changed for each items + filter on them allows clients to only request the changed data and sync that changed data and not all the database.

**Reply by dbrader** [reply]
> Having a client sync 300 000 songs just to get a dozens of last played or 2 favorites changes is not good.

That's not how REST sync works.  And I never suggested anything like that.  It typically has lists of entity types that contain change flags to determine if the client needs to sync a given entity.  If there are too many entity types, the api would contain an endpoint that is a list of lists.  Frequently changed data would be separated out in the api, or simply included in the list itself, to reduce overhead.  For example, the api would have a favorite song list so the client would not have to process all 300,000 songs to get the changes.  I also said getting a complete song list might be a good choice for differential sync because of the high number of entities.  But do we really need this level of overhead for playlists, artists or favorite albums?

Having an api that filters the data changes causes the remote client to process that list then iterate through the differential sync list to detect deletions.  In the case of playlists, forcing the client to make at least 2 network calls then iterate through 2 lists when the data would most certainly fit in one mtu is not good either.

My issue is does the api move towards a more RESTful approach or this un-RESTful hybrid approach that *could* actually take longer to sync, is harder to implement and generates more bugs.

**Reply by Tolriq** [reply]
You are again mixing a tons of different things and concept here .... The API is already mostly RESTful even if not using the HTTP verbs to handle modification and deletions.

**You seems to only think about playlists and groups of things like albums.**

I'm talking about proper offline first app with complete data sync ....

There's no concept of favorite list or anything in that case, the app needs to have a copy of the 300 000 songs. To update the non breaking changes filtering by last changed is the only efficient in a rest way without a database log system ....

For the rest sorry most also does not make sense about how things currently work. The playlist changed attribute only refer to changes to the actual playlist, so name or image or song list content / order change. If any song inside the playlist is changed like increased playcount or changed name or metadata then changed of the playlist is not changed at all.

So in all cases to have the up to date data you already need multiple calls... Because if a playlist changed have not changed it does not mean the songs have not changed their data.

**Comment by gravelld** [comment]
What does a change mean for non-atomic entities? For an album, has it changed if:
- One of its tracks' audio has changed? 
- One of its tracks' metadata has changed?
- Derived metadata, e.g. play counts, have changed?

Same for, say, genres, considering constituent albums and tracks?

**Reply by Tolriq** [reply]
It would mean any property that is exposed for that entity. So there's also the scraped metadata or the cover art. And the number of songs if there's addition / deletion or replacement of some.

Changes that don't impact returned data for an album like song metadata / user data change should not change this else clients would resync the album for no reason as it should already have detected the song changes and synced them.
